File: Solo Practice/Codeforces/unsolved/743D.py
import sys
sys.setrecursionlimit(1<<25)
INF=1<<60

def dfs(w, e, u=0, p=-1):
	ans, x, y=[-INF]*3
	s=w[u]
	for v in e[u]:
		if v==p:
			continue
		a, m, t=dfs(w, e, v, u)
		ans=max(ans, a)
		s+=t
		if y<m:
			y=m
		if x<y:
			x, y=y, x
	if y>-INF:
		ans=max(ans, x+y)
	return ans, max(x, s), s

n=int(input())
e=[[] for _ in range(n)]
w=list(map(int, input().split()))
for i in range(1, n):
	u, v=map(int, input().split())
	u, v=u-1, v-1
	e[u]+=[v]
	e[v]+=[u]
ans=dfs(w, e)[0]
print('Impossible' if ans<=-INF else ans)

File: Solo Practice/Codeforces/AC/743D.cpp
#include<cstdio>
#include<vector>
#include<algorithm>
using namespace std;
typedef long long lld;
const lld INF=1LL<<60;

struct data
{
    lld a, m, s;
};

data dfs(const vector<int>& w, const vector<vector<int>>& e, int u=1, int p=0)
{
    lld a=-INF, x=-INF, y=-INF, s=w[u];
    
    for(const int v : e[u])
    {
        if( v==p )
            continue;
        
        data tmp=dfs(w, e, v, u);
        a=max(a, tmp.a);
        s+=tmp.s;
        
        if( y<tmp.m )
            y=tmp.m;
        
        if( x<y )
            swap(x, y);
    }
    
    if( y>-INF )
        a=max(a, x+y);
    
    return {a, max(x, s), s};
}

int main()
{
    int n;
    scanf("%d", &n);
    vector<vector<int>> e(n+1);
    vector<int> w(n+1, 0);
    
    for(int i=1; i<=n; i++)
        scanf("%d", &w[i]);
    
    for(int i=1; i<n; i++)
    {
        int u, v;
        scanf("%d%d", &u, &v);
        e[u].push_back(v);
        e[v].push_back(u);
    }
    
    lld ans=dfs(w, e).a;
    
    if( ans<=-INF )
        printf("Impossible\n");
    else
        printf("%lld\n", ans);
}